cookbooks/fb_networkmanager/recipes/default.rb

```diff
@@ -0,0 +1,104 @@
+#
+# Cookbook:: fb_networkmanager
+# Recipe:: default
+#
+# Copyright (c) 2020-present, Vicarious, Inc.
+# Copyright (c) 2020-present, Facebook, Inc.
+# All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+vpnhome = '/var/lib/openvpn/chroot'
+FB::Users.initialize_group(node, 'nm-openvpn')
+node.default['fb_users']['users']['nm-openvpn'] = {
+  'gid' => 'nm-openvpn',
+  # If /var/lib/openvpn doesn't exist yet, we can't create it in time
+  # but we want to create the user, so for the first run, create it
+  # with /tmp
+  'home' => ::File.directory?(vpnhome) ? vpnhome : '/tmp',
+  'shell' => '/usr/sbin/nologin',
+  'action' => :add,
+}
+
+packages = %w{
+  network-manager
+  openvpn
+  network-manager-openvpn-gnome
+}
```

Comment:
Is it assured we definitely want OpenVPN always configured? I'd prefer making VPN optional and allow choosing either/both of at least OpenVPN and openconnect.

Reply:
This doesn't force anyone to use openvpn. It just happens to setup openvpn only because I didn't have a test-case for openconnect. It's not mutually-exclusive afaik?

Reply:
Ah, good catch @jgoguen. Yes, we should gate this behind an attribute and make it optional. VPN setup is totally fine here, but we shouldn't do it by default, and we should write in a way that can support multiple backends (even if you only implement openvpn for now).

Reply:
I'm just installing the right packages, you can make NM use whatever backened you want in the connection. We can install all the backends. This doesn't configure any VPNs... it just makes it available.

Reply:
But if I'm never going to use OpenVPN, I need `openconnect`, there's no point to having it installed. If we gate it behind an attribute we can allow administrators to better control what gets installed.

As soon as we're going to have multiple VPN backends installed by Chef (plus any sanity work, like making sure the user is created) it may be better if we instead have a series of includes:

```
include_recipe 'fb_networkmanager::openvpn
include_recipe 'fb_networkmanager::openconnect
```

Each one can have an early return based on something like `node['fb_networkmanager']['enable'] && node['fb_networkmanager'][vpntype]['enable']`

Reply:
No, they can't. That's not runtime safe.

If you're just worried about packages, set `manage_packages` to `false`, and install whatever you want.

In general, my take on this is "install the extra plugins so users can add whatever connections they want" - the plugins are small, who cares?

Reply:
Oh, yes you're right. It would need to be a set of resources (or a resource called multiple times with different actions, which just no).

@davide125 any strong opinion? I'd still like to have some gate for these but not strongly enough to demand that much extra work and testing for little gain.

Reply:
You have a gate, `manage_packages`. :)

Reply:
I'm not convinced we should be installing `openvpn` (and `openconnect` or `vpnc` or whatever for that matter) itself here; a bunch of these will happily drop off system services, and can have interesting side effects if not setup properly. I think the ideal state would be that we have `fb_openvpn` / `fb_openconnect` / etc to manage them properly, and that `fb_networkmanager` just uses their API and takes care of the n-m specific bits (in this case, the user and the plugin package). I do realize this is a fuckton of work though, and it doesn't make sense to block this PR on it.

How about this: we move the openvpn stuff to a `fb_networkmanager::openvpn`, and we just don't include it in the default recipe. In the README, we document that if you want openvpn you should include. This is akin to how we handle some of the more corner cases in `fb_systemd` iirc.

Reply:
except that doesn't make sense for fb_networkmanater to use an fb_openvpn API, because it's not configuring it statically. NetworkManager *internally* uses those things, but it needs the packages installed.

We can certainly make extra recipes for them and that's fine, but it's literally Just The Packages, to be clear.

Reply:
I still contend we should split it out. If we had a `fb_openvpn`, the way I see it `fb_networkmanager::openvpn` would just include `fb_openvpn::packages` and then install its thing. For now, I think it'd be fine to install both `openvpn` and the other package directly in a `fb_networkmanager::openvpn`. N-M can support like 5 or 6 different type of VPNs, some with fairly complex and invasive dependencies. We shouldn't force the user to have them all installed by default.